src/ray/core_worker/store_provider/memory_store/memory_store.cc

```diff
@@ -358,10 +362,19 @@ Status CoreWorkerMemoryStore::Wait(const absl::flat_hash_set<ObjectID> &object_i
   return Status::OK();
 }
 
-void CoreWorkerMemoryStore::Delete(const absl::flat_hash_set<ObjectID> &object_ids) {
+void CoreWorkerMemoryStore::Delete(const absl::flat_hash_set<ObjectID> &object_ids,
+                                   absl::flat_hash_set<ObjectID> *plasma_ids_to_delete) {
   absl::MutexLock lock(&mu_);
   for (const auto &object_id : object_ids) {
-    objects_.erase(object_id);
+    auto it = objects_.find(object_id);
+    if (it != objects_.end()) {
+      if (it->second->IsInPlasmaError()) {
+        plasma_ids_to_delete->insert(
+            object_id.WithTransportType(TaskTransportType::RAYLET));
+      } else {
```

Comment:
This seems awkward, how about we don't do this and the higher level delete call can just go to both stores (i.e., always try to delete id.WithTransportType(RAYLET)). It will be a bit slower, but I don't think delete performance is important.

Reply:
We still need something like this so that we don't accidentally delete the IsInPlasmaError. Otherwise, the caller will just hang forever because it won't be able to find the in-memory store value.

The free call is actually quite problematic now that we're using the owner to track eviction errors. If the user is allowed to call free randomly, then we'd have to contact the owner to check for eviction for every object ID, whether it was serialized or not, to prevent hangs. I'm not quite sure how to fix this behavior.

Reply:
Hmm yeah, I guess delete() bypasses ref counting in a bad way.